R/S3.R

```diff
@@ -241,3 +313,36 @@ class_data.frame <- new_S3_class("data.frame",
   },
   validator = validate_data.frame
 )
+
+#' @export
+#' @rdname base_s3_classes
+#' @format NULL
+#' @order 3
+class_matrix <- new_S3_class("matrix",
+  constructor = function(.data = NA, nrow = 1, ncol = 1, byrow = FALSE, dimnames  = NULL) {
+    matrix(.data, nrow, ncol, byrow, dimnames)
+  },
+  validator = validate_matrix
+)
+
+#' @export
+#' @rdname base_s3_classes
+#' @format NULL
+#' @order 3
+class_array <- new_S3_class("array",
+  constructor = function(.data = NA, dim = length(data), dimnames = NULL) {
+    array(.data, dim, dimnames)
+  },
+  validator = validate_array
+)
+
+#' @export
+#' @rdname base_s3_classes
+#' @format NULL
+#' @order 3
+class_formula <- new_S3_class("formula",
+  constructor = function(.data = NULL, env = parent.frame()) {
+    stats::formula(.data, env)
```

Comment:
`stats::formula()` is itself an S3 generic. Should the default constructor for `class_formula` dispatch a generic? This breaks the pattern established with other classes we could choose to dispatch for, but don't,, such as `POSIXct`, `factor`, `array`, etc.

Reply:
I don't see a problem with it. It's the only abstraction available. The other weird thing about formula is that it is the only "base" class not defined by the base package.